test/e2e/controller_test.go

```diff
@@ -0,0 +1,230 @@
+		},
+		Spec: v1alpha1.TemplateSpec{
+			Body:       "---\nFIRSTNAME: foo\nLASTNAME: bar",
+			Parameters: []string{"FIRSTNAME", "LASTNAME"},
+		},
+	}
+
+	validCombinationCR := v1alpha1.Combination{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "validcombination",
+		},
+		Spec: v1alpha1.CombinationSpec{
+			Arguments: []v1alpha1.Argument{
+				{
+					Key: "FIRSTNAME",
+					Values: []string{
+						"John",
+						"Luke",
+					},
+				},
+				{
+					Key: "LASTNAME",
+					Values: []string{
+						"Snow",
+						"Skywalker",
+					},
+				},
+			},
+		},
+	}
+
+	expectedEvaluations := []string{
+		"John: Snow",
+		"John: Skywalker",
+		"Luke: Snow",
+		"Luke: Skywalker",
+	}
+
+	expectedUpdatedEvaluations := []string{
+		"John: foo\nSnow: bar",
+		"John: foo\nSkywalker: bar",
+		"Luke: foo\nSkywalker: bar",
+		"Luke: foo\nSnow: bar",
+	}
+
+	When("given healthy input and a healthy template", func() {
+		var ctx context.Context
+		var validTemplateCRCopy *v1alpha1.Template
+		var validCombinationCRCopy *v1alpha1.Combination
+
+		BeforeEach(func() {
+			ctx = context.Background()
+
+			// Create copies of valid CR's
+			validTemplateCRCopy = validTemplateCR.DeepCopy()
+			validCombinationCRCopy = validCombinationCR.DeepCopy()
+
+			// Create a valid template
+			err := kubeclient.Create(ctx, validTemplateCRCopy)
+			Expect(err).To(BeNil(), "failed to create template CR")
+
+			// Create a valid combination
+			validCombinationCRCopy.Spec.Template = validTemplateCRCopy.Name
+			err = kubeclient.Create(ctx, validCombinationCRCopy)
+			Expect(err).To(BeNil(), "failed to create combination CR")
+		})
+
+		AfterEach(func() {
+			err := kubeclient.Delete(ctx, validCombinationCRCopy)
+			Expect(err).To(BeNil(), "failed to clean-up combination CR after test")
+
+			err = kubeclient.Delete(ctx, validTemplateCRCopy)
+			Expect(err).To(BeNil(), "failed to clean-up template CR after test")
+
+			ctx.Done()
+		})
+
+		It("should get the correct evaluations and a Processed status", func() {
+			// Give up to a minute (default eventually timeout) for the combination to process properly
+			Eventually(func(g Gomega) error {
+				var retrievedCombination v1alpha1.Combination
+				err := kubeclient.Get(ctx, types.NamespacedName{Name: validCombinationCRCopy.Name}, &retrievedCombination)
+
+				var conditionReasons []string
+				for _, condition := range retrievedCombination.Status.Conditions {
+					conditionReasons = append(conditionReasons, condition.Reason)
+				}
+
+				g.Expect(conditionReasons).To(ContainElement(v1alpha1.ReasonProcessed))
+				g.Expect(retrievedCombination.Status.Evaluations).To(ContainElements(expectedEvaluations))
+
+				return err
+			}).Should(BeZero())
+		})
+
+		It("should reevaluate whenever its associated template gets updated", func() {
+			// Update template to trigger a requeue event
+			validTemplateCRCopy.Spec.Body = validUpdatedTemplateCR.Spec.Body
+			err := kubeclient.Update(ctx, validTemplateCRCopy)
```

Comment:
(not blocking) Wrap this update call in an Eventually handler to cut down on flakes in the long run? You may need to also perform a GET operation first on this resource, then an update (in the same eventually block).